File: src/endpoints/category/put_item.py
from . import __types__ as types
from src import output
from src.libs.db import DB, Table
from src.modules.verify import checking_token
from . import __libs__ as category_libs

async def put_item(params: dict, req = None, _db: DB = None, _check_token = True):

    # set values
    result = None
    db = _db if _db else DB().connect()

    try:
        # set params
        params = types.PutItem(**params)

        # checking token
        if _check_token: checking_token(req, db)

        # check module
        category_libs.check_module(db, params.module, params.module_srl)

        # check module and get max turn
        where = [ f'AND module LIKE \'{params.module}\'' ]
        match params.module:
            case category_libs.Module.NEST:
                if params.module_srl:
                    where.append(f'AND module_srl = {params.module_srl}')
                else:
                    where.append(f'AND module_srl IS NULL')
            case category_libs.Module.JSON:
                where.append(f'AND module_srl IS NULL')
        count = db.get_count(
            table_name=Table.CATEGORY.value,
            where=where,
        )

        # set values
        values = {
            'name': params.name,
            'module': params.module,
            'turn': count + 1,
        }
        match params.module:
            case category_libs.Module.NEST:
                if params.module_srl: values['module_srl'] = params.module_srl

        # set placeholders
        placeholders = [
            { 'key': 'name', 'value': ':name' },
            { 'key': 'module', 'value': ':module' },
        ]
        if 'module_srl' in values:
            placeholders.append({ 'key': 'module_srl', 'value': ':module_srl' })
        placeholders.append({ 'key': 'turn', 'value': ':turn' })
        placeholders.append({ 'key': 'created_at', 'value': 'DATETIME("now", "localtime")' })

        # add item
        data = db.add_item(
            table_name=Table.CATEGORY.value,
            placeholders=placeholders,
            values=values,
        )

        # set result
        result = output.success({
            'message': 'Complete add category.',
            'data': data,
        }, _req=req)
    except Exception as e:
        result = output.exc(e, _req=req)
    finally:
        if not _db and db: db.disconnect()
        return result

File: src/routes/category/put-item.ts
import DB, { db } from '@/classes/DB'
import ServiceError from '@/classes/ServiceError'
import * as helper from './__helper'
import type { CategoryModel } from './__model'

type PutItemParams = {
  body: CategoryModel['putItemBody'],
}

export default async function putItem({ body }: PutItemParams)
{
  try
  {
    // checking module
    helper.checkingModule(body.module, body.module_srl)

    // checking exist name
    helper.checkingExistName({
      module: body.module,
      moduleSrl: body.module_srl,
      name: body.name,
    })

    // get count for max turn
    let _where = [ `AND module LIKE \'${body.module}\'` ]
    switch (body.module)
    {
      case helper.MODULE.NEST:
        if (body.module_srl) _where.push(`AND module_srl = ${body.module_srl}`)
        else _where.push(`AND module_srl IS NULL`)
        break
      case helper.MODULE.JSON:
        _where.push(`AND module_srl IS NULL`)
        break
    }
    const count = db.getCount({
      table: DB.TABLE.CATEGORY,
      where: _where,
    })

    // add data
    const added = db.addData({
      table: DB.TABLE.CATEGORY,
      values: [
        { key: 'name', value: body.name },
        { key: 'turn', value: count.data + 1 },
        { key: 'module', value: body.module },
        (body.module === helper.MODULE.NEST && body.module_srl) && { key: 'module_srl', value: body.module_srl },
        { key: 'created_at', valueName: DB.DATE_TIME },
      ],
    })

    return added.data
  }
  catch (_e: any)
  {
    throw new ServiceError('Failed to add Category.', {
      status: _e.status,
      text: _e.message,
      cause: _e,
    })
  }
}
